test/refine_struct.lean

```diff
@@ -1,6 +1,53 @@
 import tactic.interactive
 
-section refine_struct
+/-!
+`refine_struct` caused a variety of interesting problems,
+which were identified in
+https://github.com/leanprover-community/mathlib/pull/2251
+and
+https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Need.20help.20with.20class.20instance.20resolution
+
+These tests are quite specific to testing the patch made in
+https://github.com/leanprover-community/mathlib/pull/2319
+and are not a complete test suite for `refine_struct`.
+-/
+
+instance pi_has_one {α : Type*} {β : α → Type*} [Π x, has_one (β x)] : has_one (Π x, β x) :=
+by refine_struct { .. }; exact λ _, 1
+
+open tactic
+
+run_cmd (do
+  (declaration.defn _ _ _ b _ _) ← resolve_constant `pi_has_one >>= get_decl,
+  -- Make sure that `eq.mpr` really doesn't occur in the body:
+  eq_mpr ← mk_const `eq.mpr,
+  k ← kabstract b eq_mpr, -- `expr.occurs` doesn't work here, always giving `ff` even before the patch.
+  guard (k = b) <|>
+    fail "result generated by `refine_struct` contained an unnecessary `eq.mpr`",
+  -- Make sure that `id` really doesn't occur in the body:
+  id ← mk_const `id,
+  k ← kabstract b id,
+  guard (k = b) <|>
+    fail "result generated by `refine_struct` contained an unnecessary `id`")
+
+-- Next we check that fields defined for embedded structures are unfolded
+-- when seen by fields in the outer structure.
+structure foo (α : Type):=
+(a : α)
+
+structure bar (α : Type) extends foo α :=
+(b : a = a)
+
+example : bar ℕ :=
+begin
+  refine_struct { a := 1, .. },
+  -- We're making sure that the goal is
+  -- ⊢ 1 = 1
+  -- rather than
+  -- ⊢ {a := 1}.a = {a := 1}.a
+  target >>= pp >>= (λ t, guard $ to_string t = "1 = 1"),
```

Comment:
```suggestion
  guard_target 1 = 1,
```

Reply:
@semorrison Did you have any particular reason for this long line?

Reply:
Somehow that day `guard_target` just wasn't working for me. :-( I guess I should have stopped to work out why before trying a workaround. If it passes, great.

Reply:
I didn't try it first.  But yes it fails because the target is actually *different* from `1 = 1`:
```
@eq.{1} (@(λ {α : Type} (c : foo α), α) nat (@foo.mk nat (@has_one.one.{0} nat nat.has_one)))
    (@has_one.one.{0} nat nat.has_one)
    (@has_one.one.{0} nat nat.has_one)
```

If this beta-redex is acceptable, then I'd still use the following over the string comparison:
```lean
  (do tgt ← target, is_def_eq tgt `(1 = 1) transparency.none),
```

Reply:
@gebner, is this a problem? I'm actually not sure.

Reply:
I feel like this will cause weird problems down the line.  What was the reason to use `unfold` in the first place?  If I replace `interactive.dunfold ....` with the equivalent of `dsimp only` then all the tests pass:
```lean
             dsimp_target simp_lemmas.mk,
```

Reply:
This seems great. We don't even need the before/after `target` comparison. (I don't know the original reason to use `unfold`: this is @cipher1024's from a while back.)

Reply:
I've pushed a change with this suggestion.

Reply:
(Including the change to the test from this thread.)